conjure-go-client/httpclient/authn_test.go

```diff
@@ -113,5 +116,205 @@ func TestRoundTripperWithBasicAuthProvider(t *testing.T) {
+		},
+		{
+			Name:         "WithBasicAuth param",
+			Server:       basicAuthServer,
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithBasicAuth(username, password)},
+		},
+		{
+			Name:   "WithBasicAuthProvider param",
+			Server: basicAuthServer,
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithBasicAuthProvider(func(ctx context.Context) (httpclient.BasicAuth, error) {
+				return httpclient.BasicAuth{User: username, Password: password}, nil
+			})},
+		},
+		{
+			Name:   "WithBasicAuthOptionalProvider present",
+			Server: basicAuthServer,
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithBasicAuthOptionalProvider(func(ctx context.Context) (*httpclient.BasicAuth, error) {
+				return &httpclient.BasicAuth{User: username, Password: password}, nil
+			})},
+		},
+		{
+			Name:   "WithBasicAuthOptionalProvider absent",
+			Server: noAuthServer,
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithBasicAuthOptionalProvider(func(ctx context.Context) (*httpclient.BasicAuth, error) {
+				return nil, nil
+			})},
+		},
+		{
+			Name:         "WithBasicAuth param beats basic config",
+			Server:       basicAuthServer,
+			Config:       httpclient.ClientConfig{BasicAuth: &httpclient.BasicAuth{User: "wrong", Password: "wrong"}},
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithBasicAuth(username, password)},
+		},
+		{
+			Name:         "WithBasicAuth param beats bearer config",
+			Server:       basicAuthServer,
+			Config:       httpclient.ClientConfig{APIToken: &token},
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithBasicAuth(username, password)},
+		},
+		// Bearer tokens
+		{
+			Name:   "APIToken config",
+			Server: bearerAuthServer,
+			Config: httpclient.ClientConfig{APIToken: &token},
+		},
+		{
+			Name:   "APITokenFile config",
+			Server: bearerAuthServer,
+			Config: httpclient.ClientConfig{APITokenFile: &apiTokenFile},
+		},
+		{
+			Name:         "WithAuthToken param",
+			Server:       bearerAuthServer,
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithAuthToken(token)},
+		},
+		{
+			Name:         "WithAuthToken param beats basic config",
+			Server:       bearerAuthServer,
+			Config:       httpclient.ClientConfig{BasicAuth: &httpclient.BasicAuth{User: "wrong", Password: "wrong"}},
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithAuthToken(token)},
+		},
+		{
+			Name:   "WithAuthTokenProvider param",
+			Server: bearerAuthServer,
+			ClientParams: []httpclient.ClientOrHTTPClientParam{httpclient.WithAuthTokenProvider(func(ctx context.Context) (string, error) {
+				return token, nil
+			})},
+		},
+		// WithRequest*
+		{
+			Name:          "WithRequestBasicAuth param",
+			Server:        basicAuthServer,
+			RequestParams: []httpclient.RequestParam{httpclient.WithRequestBasicAuth(username, password)},
+		},
+		{
```

Comment:
these four tests fail on develop because the header set by WithRequestBasicAuth is overwritten by the client's middleware